src/tools/miri/tests/native-lib/pass/ptr_write_access.rs

```diff
@@ -5,118 +5,192 @@
+    }
+    unsafe impl Sync for SyncPtr {}
+
     extern "C" {
-        fn init_interior_mutable(pptr: *const UnsafeInterior);
+        fn init_static_inner(s_ptr: *const SyncPtr, val: i32);
     }
 
-    #[repr(C)]
-    struct UnsafeInterior {
-        mut_ptr: *mut i32
+    static mut INNER: MaybeUninit<i32> = MaybeUninit::uninit();
+    #[allow(static_mut_refs)]
+    static STATIC: SyncPtr = SyncPtr { ptr: unsafe { INNER.as_mut_ptr() } };
+    let val = 41;
+
+    let inner = unsafe {
+        init_static_inner(&STATIC, val);
+        INNER.assume_init()
+    };
+    assert_eq!(inner, val);
+}
+
+// Test function that writes a pointer and exposes the alloc of its int argument.
+fn test_expose_int() {
+    extern "C" {
+        fn expose_int(int_ptr: *const i32, pptr: *mut *const i32);
     }
-    unsafe impl Sync for UnsafeInterior {}
-    
-    let mut x = MaybeUninit::<i32>::uninit();
-    let unsafe_interior = UnsafeInterior { mut_ptr: x.as_mut_ptr() };
-    unsafe { init_interior_mutable(&unsafe_interior) };
 
-    assert_eq!(unsafe { x.assume_init() }, 51);
+    let x = 51;
+    let mut ptr = std::ptr::null();
+
+    unsafe { expose_int(&x, &mut ptr) };
+    assert_eq!(unsafe { *ptr }, x);
 }
 
+// Test function that swaps two pointers and exposes the alloc of an int.
 fn test_swap_ptr() {
     extern "C" {
         fn swap_ptr(pptr0: *mut *const i32, pptr1: *mut *const i32);
     }
 
-    let x = 41;
-    let mut ptr0 = &raw const x;
-    let mut ptr1 = std::ptr::null();
+    let x = 61;
+    let (mut ptr0, mut ptr1) = (&raw const x, null());
+
     unsafe { swap_ptr(&mut ptr0, &mut ptr1) };
+    assert_eq!(unsafe { *ptr1 }, x);
+}
+
+// Test function that swaps two nested pointers and exposes the alloc of an int.
+fn test_swap_nested_ptr() {
+    extern "C" {
+        fn swap_nested_ptr(ppptr0: *mut *mut *const i32, ppptr1: *mut *mut *const i32);
+    }
+
+    let x = 71;
+    let (mut ptr0, mut ptr1) = (&raw const x, null());
+    let (mut pptr0, mut pptr1) = (&raw mut ptr0, &raw mut ptr1);
 
+    unsafe { swap_nested_ptr(&mut pptr0, &mut pptr1) }
     assert_eq!(unsafe { *ptr1 }, x);
 }
 
-fn test_interact_dangling() {
+// Test function that swaps two pointers in a struct and exposes the alloc of an int.
+fn test_swap_tuple() {
+    #[repr(C)]
+    struct Tuple {
+        ptr0: *const i32,
+        ptr1: *const i32,
+    }
+
+    extern "C" {
+        fn swap_tuple(t_ptr: *mut Tuple);
+    }
+
+    let x = 81;
+    let mut tuple = Tuple { ptr0: &raw const x, ptr1: null() };
+
+    unsafe { swap_tuple(&mut tuple) }
+    assert_eq!(unsafe { *tuple.ptr1 }, x);
+}
+
+// Test function that interacts with a dangling pointer.
+fn test_overwrite_dangling() {
     extern "C" {
         fn overwrite_ptr(pptr: *mut *const i32);
     }
 
-    let x = Box::new(61);
-    let mut ptr = Box::as_ptr(&x);
-    drop(x);
+    let b = Box::new(91);
+    let mut ptr = Box::as_ptr(&b);
+    drop(b);
     unsafe { overwrite_ptr(&mut ptr) };
```

Comment:
Do we have a test where the pointer that we pass to C itself is dangling? Obviously C then can't read or write from it. But at least this shouldn't crash Miri.

That wold be like passing `ptr` here (not `&mut ptr`).